contextual_lenses/train_utils.py

```diff
@@ -34,18 +39,46 @@ def create_data_iterator(df, input_col, output_col, batch_size, epochs=1, buffer
 
   outputs = df[output_col].values
   outputs = tf.data.Dataset.from_tensor_slices(outputs)
+  
+  if add_outputs:
+    batches = tf.data.Dataset.zip((inputs, outputs)).shuffle(buffer_size=buffer_size, seed=seed)
+  else:
+    batches = inputs.shuffle(buffer_size=buffer_size, seed=seed)
+  
+  batches = batches.repeat(epochs).batch(batch_size=batch_size, drop_remainder=drop_remainder)
+    
+  if as_numpy:
+    batches = batches.as_numpy_iterator()
+    
+  return batches
 
-  batches = tf.data.Dataset.zip((inputs, outputs)).shuffle(buffer_size=buffer_size, seed=seed)
-  batches = batches.repeat(epochs).batch(batch_size=batch_size, drop_remainder=drop_remainder).as_numpy_iterator()
 
-  return batches
+def path_inclusion_filter_fn(path, param, layer):
+  """Returns whether or not layer name is contained in path."""
 
+  return layer in path
 
-def create_optimizer(model, learning_rate, weight_decay):
-  """Instantiates Adam optimizer."""
 
-  optimizer_def = optim.Adam(learning_rate=learning_rate, weight_decay=weight_decay)
-  optimizer = optimizer_def.create(model)
+def create_optimizer(model, learning_rate, weight_decay, layers=None):
```

Comment:
Lots of flax details here. make sure ddohan signs off on this entire file. Also need to have some loading/saving tests. 
@dmrd